### DISCUSSION TITLE
efficient way to do per-row rank fold

### DISCUSSION THREAD

**Discussion by zoranpopovic** [libvips/libvips#2017] [Ideas]
I'm trying to compute max intensity for each row of a large image file.  right now I do it with a rank full-width rank filter over the whole image which is wasteful. 
```
img = img.rank(img.width, 1, img.width-1)
col_mip = img.extract_area(img.width/2, 0, 1, img.height)
```

Any idea how to do this more efficiently.

**Comment by jcupitt** [comment]
Yes, that will be a bit slow. Sorry, I can't think of an efficient way to do it. 

There's `project`, which gives you row and column sums. I think you'd need to add a new parameter to that to specify the projection operator, perhaps `VipsCombine`.

https://jcupitt.github.io/libvips/API/8.7/libvips-arithmetic.html#vips-project

http://jcupitt.github.io/libvips/API/current/libvips-convolution.html#VipsCombine

**Comment by hraftery** [comment]
Is a `crop` + `max` loop better, or does it boil down to the same thing?

OP seems to be using Python? Example for Python:

```
mip = [img.crop(0, row, img.width, 1).max() for row in range(img.height)]
```

**Reply by jcupitt** [reply]
Yes, crop + max might be faster, you'd need to benchmark it.

**Reply by hraftery** [reply]
I found it approximately just as slow as the rank option so gave up on it. For my purposes, `.shrink()` then `.tolist()` and then `[max(row) for row in img_list]` is adequate and is at least a few orders of magnitude faster. Pity, because `project()` is lightning fast.

I'm also several versions behind the latest so take my findings with a grain of salt. If I should expect otherwise that would be good to know.

**Reply by jcupitt** [reply]
I made an enhancement issue for this: https://github.com/libvips/libvips/issues/5113

The API would be eg.:

```python
rows, columns = image.project(combine="max")
```

And you'd get a W x 1 and a 1 x H image containing the peak value for each row and column.

**Reply by jcupitt** [reply]
I made a PR, let's move any discussion there. https://github.com/libvips/libvips/pull/5115
